<!DOCTYPE html>
<html>
<head>
    <meta http-equiv="content-type" content="text/html; charset=UTF-8">
    <title>.NET Platform Architecture</title>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
</head>
<body>
    <style>
        body {
    padding: 60px 60px 60px 100px;
    font-family: PF Regal,Georgia,serif;
    font-size: 16pt;
    font-weight: 300;
}

strong {
    font-weight: 500;
}

p, li {
    margin: 0 0 14px;
    line-height: 32px;
    font-family: inherit;
}

ol, ul {
    padding-left: 35px;
    margin-top: 1.1rem;
    margin-bottom: 1.1rem;
}

ol li {
    padding-left: 10px;	
    font-family: inherit;
}

ul li {
    padding-left: 5px;	
    font-family: inherit;
}

code, .lang-csharp span, .lang-csharp pre {
    font-family: JB Mono, monospace !important;
}

.lang-csharp {
    padding: 4px 0px 14px 0px;
}


.lang-csharp > div {
    border-left-color: #eee;
    border-left-style: dotted;
    border-left-width: 3px;
    background-color: #f5faff !important;
    padding: 10px 10px 10px 25px;
}

code {
    font-size: 14px;
    border: gainsboro;
    border-width: 1px;
    border-style: solid;
    padding: 2px;
    color: darkslateblue;
}

.lang-csharp span, .lang-csharp pre {
    font-size: 13px;
    line-height: 20px;
}

p a, li a {
    text-decoration: none;
    color: inherit;
    box-shadow: inset 0 -1px #b88b59;
}

p img {
    width: 95%;
}

p (:has img) {
    text-align: center;
}

blockquote {
    background-color: lavenderblush;
    padding-left: 20px;
    border-left: 1px solid rgba(0,0,0,.08);
}

blockquote p {
    margin: 16px 0;
    padding: 12px 12px 13px;
    font-family: inherit;
}

blockquote p code {
    font-family: consolas, monospace;
    font-size: 17px;
    background-color: aliceblue;
    padding: 4px 6px;
}

h1, h2, h3, h4, h5, h6 {
    margin: 0;
    padding: 0;
    font-weight: 400;
    font-family: inherit;
}

h1 {
    font-size: 40px;
    font-family: PF Regal,Georgia,serif;
    font-weight: 300;
    margin: 40px 0 20px 0;
    border-bottom-style: ridge;
    padding-bottom: 10px;
}

h1 b {
    font-family: Proxima Nova,Arial,Helvetica Neue,sans-serif;
    font-weight: 700;
    font-size: 37px;
    line-height: 42px;
}

h2 {
    margin-top: 35px;
    margin-bottom: 10px;
    font-family: Proxima Nova,Arial,Helvetica Neue,sans-serif;
    font-size: 20px;
    font-weight: 700;
    line-height: 24px;
    border-bottom-color: grey;
    border-bottom-style: solid;
    padding-bottom: 5px;
}

h3 {
    margin: 30px 0 10px 0;
    font-size: 17pt;
    font-weight: 500;
}

h4 {
    margin: 20px 0 10px 0;
    font-size: 16pt;
    font-weight: 500;
    font-style: italic;
}

/* Fonts 

@font-face {
    font-family: 'PF Regal';
    src: url('../../../res/fonts/PFRegalTextPro-RegularA.eot');
    src: local('../../../res/fonts/PFRegalTextPro-RegularA'),
        url('../../../res/fonts/PFRegalTextPro-RegularA.woff2') format('woff2'),
        url('../../../res/fonts/PFRegalTextPro-RegularA.woff') format('woff'),
        url('../../../res/fonts/PFRegalTextPro-RegularA.ttf') format('truetype');
    font-weight: 300;
    font-style: normal;
}

@font-face {
    font-family: 'PF Regal';
    src: url('../../../res/fonts/PFRegalTextPro-Bold.eot');
    src: local('../../../res/fonts/PFRegalTextPro-Bold'),
        url('../../../res/fonts/PFRegalTextPro-Bold.woff2') format('woff2'),
        url('../../../res/fonts/PFRegalTextPro-Bold.woff') format('woff'),
        url('../../../res/fonts/PFRegalTextPro-Bold.ttf') format('truetype');
    font-weight: bold;
    font-style: normal;
}
*/
@font-face {
    font-family: 'PF Regal';
    src: url('../../../res/fonts/PFRegalTextPro-Medium.eot');
    src: local('../../../res/fonts/PFRegalTextPro-Medium'),
        url('../../../res/fonts/PFRegalTextPro-Medium.woff2') format('woff2'),
        url('../../../res/fonts/PFRegalTextPro-Medium.woff') format('woff'),
        url('../../../res/fonts/PFRegalTextPro-Medium.ttf') format('truetype');
    font-weight: 500;
    font-style: normal;
}

@font-face {
    font-family: 'PF Regal';
    src: url('../../../res/fonts/PFRegalTextPro-BlackItalic.eot');
    src: local('../../../res/fonts/PFRegalTextPro-BlackItalic'),
        url('../../../res/fonts/PFRegalTextPro-BlackItalic.woff2') format('woff2'),
        url('../../../res/fonts/PFRegalTextPro-BlackItalic.woff') format('woff'),
        url('../../../res/fonts/PFRegalTextPro-BlackItalic.ttf') format('truetype');
    font-weight: 900;
    font-style: italic;
}

@font-face {
    font-family: 'PF Regal';
    src: url('../../../res/fonts/PFRegalTextPro-Black.eot');
    src: local('../../../res/fonts/PFRegalTextPro-Black'),
        url('../../../res/fonts/PFRegalTextPro-Black.woff2') format('woff2'),
        url('../../../res/fonts/PFRegalTextPro-Black.woff') format('woff'),
        url('../../../res/fonts/PFRegalTextPro-Black.ttf') format('truetype');
    font-weight: 900;
    font-style: normal;
}

@font-face {
    font-family: 'PF Regal UBlack';
    src: url('../../../res/fonts/PFRegalTextPro-UBlack.eot');
    src: local('../../../res/fonts/PFRegalTextPro-UBlack'),
        url('../../../res/fonts/PFRegalTextPro-UBlack.woff2') format('woff2'),
        url('../../../res/fonts/PFRegalTextPro-UBlack.woff') format('woff'),
        url('../../../res/fonts/PFRegalTextPro-UBlack.ttf') format('truetype');
    font-weight: 900;
    font-style: normal;
}

@font-face {
    font-family: 'PF Regal';
    src: url('../../../res/fonts/PFRegalTextPro-MediumItalic.eot');
    src: local('../../../res/fonts/PFRegalTextPro-MediumItalic'),
        url('../../../res/fonts/PFRegalTextPro-MediumItalic.woff2') format('woff2'),
        url('../../../res/fonts/PFRegalTextPro-MediumItalic.woff') format('woff'),
        url('../../../res/fonts/PFRegalTextPro-MediumItalic.ttf') format('truetype');
    font-weight: 500;
    font-style: italic;
}

@font-face {
    font-family: 'PF Regal';
    src: url('../../../res/fonts/PFRegalTextPro-UBlackItalic.eot');
    src: local('../../../res/fonts/PFRegalTextPro-UBlackItalic'),
        url('../../../res/fonts/PFRegalTextPro-UBlackItalic.woff2') format('woff2'),
        url('../../../res/fonts/PFRegalTextPro-UBlackItalic.woff') format('woff'),
        url('../../../res/fonts/PFRegalTextPro-UBlackItalic.ttf') format('truetype');
    font-weight: 900;
    font-style: italic;
}

@font-face {
    font-family: 'PF Regal';
    src: url('../../../res/fonts/PFRegalTextPro-RegularAItalic.eot');
    src: local('../../../res/fonts/PFRegalTextPro-RegularAItalic'),
        url('../../../res/fonts/PFRegalTextPro-RegularAItalic.woff2') format('woff2'),
        url('../../../res/fonts/PFRegalTextPro-RegularAItalic.woff') format('woff'),
        url('../../../res/fonts/PFRegalTextPro-RegularAItalic.ttf') format('truetype');
    font-weight: 300;
    font-style: italic;
}

@font-face {
    font-family: 'PF Regal';
    src: url('../../../res/fonts/PFRegalTextPro-BoldItalic.eot');
    src: local('../../../res/fonts/PFRegalTextPro-BoldItalic'),
        url('../../../res/fonts/PFRegalTextPro-BoldItalic.woff2') format('woff2'),
        url('../../../res/fonts/PFRegalTextPro-BoldItalic.woff') format('woff'),
        url('../../../res/fonts/PFRegalTextPro-BoldItalic.ttf') format('truetype');
    font-weight: bold;
    font-style: italic;
}

/*  Regular B */
@font-face {
    font-family: 'PF Regal';
    src: url('../../../res/fonts/PFRegalTextPro-RegularBItalic.eot');
    src: local('../../../res/fonts/PFRegalTextPro-RegularBItalic'),
        url('../../../res/fonts/PFRegalTextPro-RegularBItalic.woff2') format('woff2'),
        url('../../../res/fonts/PFRegalTextPro-RegularBItalic.woff') format('woff'),
        url('../../../res/fonts/PFRegalTextPro-RegularBItalic.ttf') format('truetype');
    font-weight: normal;
    font-style: italic;
}

@font-face {
    font-family: 'PF Regal';
    src: url('../../../res/fonts/PFRegalTextPro-RegularB.eot');
    src: local('../../../res/fonts/PFRegalTextPro-RegularB'),
        url('../../../res/fonts/PFRegalTextPro-RegularB.woff2') format('woff2'),
        url('../../../res/fonts/PFRegalTextPro-RegularB.woff') format('woff'),
        url('../../../res/fonts/PFRegalTextPro-RegularB.ttf') format('truetype');
    font-weight: normal;
    font-style: normal;
}                
    </style>

<div style="margin:auto; width:100%;">
<div style="width:1000px;">
    <h2 id="section">Архитектура исключительной ситуации</h2>
<blockquote>
<p><a href="https://github.com/sidristij/dotnetbook/issues/51">Ссылка на обсуждение</a></p>
</blockquote>
<p>Наверное, один из самых важных вопросов, который касается нашей темы - это вопрос построения архитектуры исключений в вашем приложении. Этот вопрос интересен по многим причинам. Как по мне так основная - это видимая простота, с которой не всегда очевидно, что делать. Это свойство присуще всем базовым конструкциям, которые используются повсеместно: это и <code>IEnumerable</code>, и <code>IDisposable</code> и <code>IObservable</code> и прочие-прочие. С одной стороны, своей простотой они манят, вовлекают в использование себя в самых разных ситуациях. А с другой стороны, они полны омутов и бродов, из которых, не зная, как иной раз и не выбраться вовсе. И, возможно, глядя на будущий объем у вас созрел вопрос: так что же такого в исключительных ситуациях?</p>
<p>Но для того чтобы прийти к каким-то выводам относительно построения архитектуры классов исключительных ситуаций мы должны с вами скопить некоторый опыт относительно их классификации. Ведь только поняв, с чем мы будем иметь дело, как и в каких ситуациях программист должен выбирать тип ошибки, а в каких - делать выбор относительно перехвата или пропуска исключений, можно понять как можно построить систему типов таким образом чтобы это стало очевидно для пользователя вашего кода. А потому, попробуем классифицировать исключительные ситуации (не сами типы исключений, а именно ситуации) по различным признакам.</p>
<h3 id="section-1">По теоретической возможности перехвата проектируемого исключения</h3>
<p>По теоретическому перехвату исключения можно легко поделить на два вида: на те, которые перехватывать будут точно и на те, которые с высокой степенью вероятности перехватывать не будут. Почему <em>с высокой степенью вероятности</em>? Потому что всегда найдётся тот, кто попробует перехватить, хотя это и не нужно было совершенно делать.</p>
<p>Давайте для начала раскроем особенности первой группы: исключения, которые должны и будут перехватывать.</p>
<p>Когда мы вводим исключение такого типа, то одной стороны мы сообщаем внешней подсистеме что мы вошли в положение, когда дальнейшие действия в рамках наших данных не имеют смысла. А с другой имеем в виду, что ничего глобального сломано не было и если нас убрать, то ничего не поменяется, а потому это исключение может быть легко перехвачено чтобы поправить ситуацию. Это свойство очень важно: именно оно определяет критичность ошибки и уверенность в том что если перехватить исключение и просто очистить ресурсы, то можно спокойно выполнять код дальше.</p>
<p>Вторая группа как бы это ни звучало странно - отвечает за исключения, которые перехватывать не нужно. Они могут быть использованы только для записи в журнал ошибок, но не для того чтобы можно было как-то поправить ситуацию. Самый простой пример - это исключения группы <code>ArgumentException</code> и <code>NullReferenceException</code>. Ведь в нормальной ситуации вы не должны, например, перехватывать исключение <code>ArgumentNullException</code> потому что источником проблемы тут будете являться именно вы, а не кто либо ещё. Если вы перехватываете данное исключение, то тем самым вы допускаете что вы ошиблись и отдали методу то, что отдавать ему было нельзя:</p>
<div class="lang-csharp editor-colors"><div style="color:Black;background-color:White;"><pre>
<span style="color:Blue;">void</span> SomeMethod(<span style="color:Blue;">object</span> argument)
{
    <span style="color:Blue;">try</span> {
        AnotherMethod(argument);
    } <span style="color:Blue;">catch</span> (ArgumentNullException exception)
    {
        <span style="color:Green;">// Log it</span>
    }
}
</pre></div>
</div>
<p>В этом методе мы пробуем перехватить <code>ArgumentNullException</code>. Но на мой взгляд его перехват выглядит очень странным: прокинуть корректные аргументы методу - полностью наша забота. Было бы не корректно среагировать постфактум: в такой ситуации самое правильное что только можно сделать - это проверить передаваемые данные заранее, до вызова метода или же что ещё лучше - построить код таким образом чтобы получение неправильных параметров было бы попросту не возможным.</p>
<p>Ещё одна группа - это исключения фатальных ошибок. Если сломан некий кэш и работа подсистемы в любом случае будет не корректной? Тогда это - фатальная ошибка и ближайший по стеку код её перехватывать гарантированно не станет:</p>
<div class="lang-csharp editor-colors"><div style="color:Black;background-color:White;"><pre>
T GetFromCacheOrCalculate()
{
    <span style="color:Blue;">try</span> {
        <span style="color:Blue;">if</span>(_cache.TryGetValue(Key, <span style="color:Blue;">out</span> <span style="color:Blue;">var</span> result))
        {
            <span style="color:Blue;">return</span> result;
        } <span style="color:Blue;">else</span> {
            T res = Strategy(Key);
            _cache[Key] = res;
            <span style="color:Blue;">return</span> res;
        }
    } cache (CacheCorreptedException exception)
    {
        RecreateCache();
        <span style="color:Blue;">return</span> GetFromCacheOrCalculate();
    }
}
</pre></div>
</div>
<p>И пусть <code>CacheCorreptedException</code> - это исключение, означающее &quot;кэш на жёстком диске не консистентен&quot;. Тогда получается, что если причина такой ошибки фатальна для подсистемы кэширования (например, отсутствуют права доступа к файлу кэша), то дальнейший код, если не сможет пересоздать кэш командой <code>RecreateCache</code>, а потому факт перехвата этого исключения является ошибкой сам по себе.</p>
<h3 id="section-2">По фактическому перехвату исключительной ситуации</h3>
<p>Ещё один вопрос, который останавливает наш полет мысли в программировании алгоритмов - это понимание: стоит ли перехватывать те или иные исключения или же стоит пропустить их сквозь себя кому-то более понимающему. Переводя на язык терминов вопрос, который нам надо решить - разграничить зоны ответственности. Давайте рассмотрим следующий код:</p>
<div class="lang-csharp editor-colors"><div style="color:Black;background-color:White;"><pre>
<span style="color:Blue;">namespace</span> JetFinance.Strategies
{
    <span style="color:Blue;">public</span> <span style="color:Blue;">class</span> WildStrategy : StrategyBase
    {
        <span style="color:Blue;">private</span> Random random =  <span style="color:Blue;">new</span> Random();

        <span style="color:Blue;">public</span> <span style="color:Blue;">void</span> PlayRussianRoulette()
        {
            <span style="color:Blue;">if</span>(DateTime.Now.Second == (random.Next() % 60))
            {
                <span style="color:Blue;">throw</span> <span style="color:Blue;">new</span> StrategyException();
            }
        }
    }

    <span style="color:Blue;">public</span> <span style="color:Blue;">class</span> StrategyException : Exception { <span style="color:Green;">/* .. */</span> }
}

<span style="color:Blue;">namespace</span> JetFinance.Investments
{
    <span style="color:Blue;">public</span> <span style="color:Blue;">class</span> WildInvestment
    {
        WildStrategy _strategy;

        <span style="color:Blue;">public</span> WildInvestment(WildStrategy strategy)
        {
            _strategy = strategy;
        }

        <span style="color:Blue;">public</span> <span style="color:Blue;">void</span> DoSomethingWild()
        {
            ?<span style="color:Blue;">try</span>?
            {
                _strategy.PlayRussianRoulette();
            }
            <span style="color:Blue;">catch</span>(StrategyException exception)
            {
            }
        }
    }
}

<span style="color:Blue;">using</span> JetFinance.Strategies;
<span style="color:Blue;">using</span> JetFinance.Investments;

<span style="color:Blue;">void</span> Main()
{
    <span style="color:Blue;">var</span> foo = <span style="color:Blue;">new</span> WildStrategy();
    <span style="color:Blue;">var</span> boo = <span style="color:Blue;">new</span> WildInvestment(foo);

    ?<span style="color:Blue;">try</span>?
    {
        boo.DoSomethingWild();
    }
    <span style="color:Blue;">catch</span>(StrategyException exception)
    {
    }
}
</pre></div>
</div>
<p>Какая из двух предложенных стратегий является более корректной? Зона ответственности - это очень важно. Изначально может показаться, что поскольку работа <code>WildInvestment</code> и его консистентность целиком и полностью зависит от <code>WildStrategy</code>, то если <code>WildInvestment</code> просто проигнорирует данное исключение, оно уйдёт в уровень повыше и делать ничего более не надо. Однако, прошу заметить что существует чисто архитектурная проблема: метод <code>Main</code> ловит исключение из архитектурно одного слоя, вызывая метод архитектурно - другого. Как это выглядит с точки зрения использования? Да, в общем, так и выглядит:</p>
<ul>
<li>заботу об этом исключении просто перевесили на нас;</li>
<li>у пользователя данного класса нет уверенности, что это исключение прокинуто через ряд методов до нас специально</li>
<li>мы начинаем тянуть лишние зависимости, от которых мы избавились, вызывая промежуточный слой.</li>
</ul>
<p>Однако, из данного вывода следует другой: <code>catch</code> мы должны ставить в методе <code>DoSomethingWild</code>. И это для нас несколько странно: <code>WildInvestment</code> вроде как жёстко зависим от кого-то. Т.е. если <code>PlayRussianRoulette</code> отработать не смог, то и <code>DoSomethingWild</code> тоже: кодов возврата тот не имеет, а сыграть в рулетку он обязан. Что же делать в такой, казалось бы, безвыходной ситуации? Ответ на самом деле прост: находясь в другом слое, <code>DoSomethingWild</code> должен выбросить собственное исключение, которое относится к этому слою и обернуть исходное как оригинальный источник проблемы - в <code>InnerException</code>:</p>
<div class="lang-csharp editor-colors"><div style="color:Black;background-color:White;"><pre>
<span style="color:Blue;">namespace</span> JetFinance.Strategies
{
    pubilc <span style="color:Blue;">class</span> WildStrategy
    {
        <span style="color:Blue;">private</span> Random random =  <span style="color:Blue;">new</span> Random();

        <span style="color:Blue;">public</span> <span style="color:Blue;">void</span> PlayRussianRoulette()
        {
            <span style="color:Blue;">if</span>(DateTime.Now.Second == (random.Next() % 60))
            {
                <span style="color:Blue;">throw</span> <span style="color:Blue;">new</span> StrategyException();
            }
        }
    }

    <span style="color:Blue;">public</span> <span style="color:Blue;">class</span> StrategyException : Exception { <span style="color:Green;">/* .. */</span> }
}

<span style="color:Blue;">namespace</span> JetFinance.Investments
{
    <span style="color:Blue;">public</span> <span style="color:Blue;">class</span> WildInvestment
    {
        WildStrategy _strategy;

        <span style="color:Blue;">public</span> WildInvestment(WildStrategy strategy)
        {
            _strategy = strategy;
        }

        <span style="color:Blue;">public</span> <span style="color:Blue;">void</span> DoSomethingWild()
        {
            <span style="color:Blue;">try</span>
            {
                _strategy.PlayRussianRoulette();
            }
            <span style="color:Blue;">catch</span>(StrategyException exception)
            {
                <span style="color:Blue;">throw</span> <span style="color:Blue;">new</span> FailedInvestmentException(<span style="color:#A31515;">&quot;Oops&quot;</span>, exception);
            }
        }
    }

    <span style="color:Blue;">public</span> <span style="color:Blue;">class</span> InvestmentException : Exception { <span style="color:Green;">/* .. */</span> }

    <span style="color:Blue;">public</span> <span style="color:Blue;">class</span> FailedInvestmentException : Exception { <span style="color:Green;">/* .. */</span> }
}

<span style="color:Blue;">using</span> JetFinance.Investments;

<span style="color:Blue;">void</span> Main()
{
    <span style="color:Blue;">var</span> foo = <span style="color:Blue;">new</span> WildStrategy();
    <span style="color:Blue;">var</span> boo = <span style="color:Blue;">new</span> WildInvestment(foo);

    <span style="color:Blue;">try</span>
    {
        boo.DoSomethingWild();
    }
    <span style="color:Blue;">catch</span>(FailedInvestmentException exception)
    {
    }
}
</pre></div>
</div>
<p>Обернув исключение другим мы по сути переводим проблематику из одного слоя приложения в другой, сделав его работу более предсказуемой с точки зрения пользователя этого класса: метода <code>Main</code>.</p>
<h3 id="section-3">По вопросам переиспользования</h3>
<p>Очень часто перед нами встаёт непростая задача: с одной стороны нам лень создавать новый тип исключения, а когда мы всё-таки решаемся, не всегда ясно от чего отталкиваться: какой тип взять за основу как базовый. А ведь именно эти решения и определяют всю архитектуру исключительных ситуаций. Давайте пробежимся по популярым решениям и сделаем некоторые выводы.</p>
<p>При выборе типа исключений можно попробовать взять уже существующее решение: найти исключение с похожим смыслом в названии и использовать его. Например, если нам отдали через параметр какую-либо сущность, которая нас почему-то не устраивает, мы можем выбросить <code>InvalidArgumentException</code>, указав причину ошибки - в Message. Этот сценарий выглядит хорошо, особенно с учётом того что <code>InvalidArgumentException</code> находится в группе исключений, которые не подлежат обязательному перехвату. Но плохим будет выбор <code>InvalidDataException</code> если вы работаете с какими-либо данными. Просто потому что этот тип находится в зоне <code>System.IO</code>, а это врядли то, чем вы занимаетесь. Т.е. получается что найти существующий тип потому что лениво делать свой - практически всегда будет не правильным подходом. Исключений, которые созданы для общего круга задач почти не существует. Практически все из них созданы под конкретные ситуации и их переиспользование будет грубым нарушением архитектуры исключительных ситуаций. Мало того, получив исключение определённого типа (например, тот же <code>System.IO.InvalidDataException</code>), пользователь будет запутан: с одной стороны он увидит источник проблемы в <code>System.IO</code> как пространство имён исключения, а с другой - совершенно другое пространство имён точки выброса. Плюс ко всему, задумавшись о правилах выброса этого исключения зайдёт на <a href="https://referencesource.microsoft.com/">referencesource.microsoft.com</a> и найдёт <a href="https://referencesource.microsoft.com/#System/sys/System/IO/compression/InvalidDataException.cs,2b389f14fb01ad1b,references">все места его выброса</a>:</p>
<ul>
<li><code>internal class System.IO.Compression.Inflater</code></li>
</ul>
<p>И поймёт что <del>просто у кого-то кривые руки</del> выбор типа исключения его запутал, поскольку метод, выбросивший исключение компрессией не занимался.</p>
<p>Также в целях упрощения переиспользования можно просто взять и создать какое-то одно исключение, объявив у него поле <code>ErrorCode</code> с кодом ошибки и жить себе припеваючи. Казалось бы: хорошее решение. Бросаете везде одно и то же исключение, выставив код, ловите всего-навсего одним <code>catch</code> повышая тем самым стабильность приложения: и делать более ничего не надо. Однако, прошу не согласиться с такой позицией. Действуя таким образом по всему приложению вы с одной стороны, конечно, упрощаете себе жизнь. Но с другой - вы отбрасываете возможность ловить подгруппу исключений, объединённых некоторой общей особенностью. Как это сделано, например, с <code>ArgumentException</code>, который под собой объединяет, целую группу исключений путём наследования. Второй серьёзный минус - чрезмерно большие и нечитаемые простыни кода, который будет организовывать фильтрацию по коду ошибки. А вот если взять другую ситуацию: когда конечному пользователю конкретизация ошибки не должна быть важна, введение обобщающего типа плюс код ошибки выглядит уже куда более правильным применением:</p>
<div class="lang-csharp editor-colors"><div style="color:Black;background-color:White;"><pre>
<span style="color:Blue;">public</span> <span style="color:Blue;">class</span> ParserException
{
    <span style="color:Blue;">public</span> ParserError ErrorCode { <span style="color:Blue;">get</span>; }

    <span style="color:Blue;">public</span> ParserException(ParserError errorCode)
    {
        ErrorCode = errorCode;
    }

    <span style="color:Blue;">public</span> <span style="color:Blue;">override</span> <span style="color:Blue;">string</span> Message
    {
        <span style="color:Blue;">get</span> {
            <span style="color:Blue;">return</span> Resources.GetResource($<span style="color:#A31515;">&quot;{nameof(ParserException)}{Enum.GetName(typeof(ParserError), ErrorCode)}&quot;</span>);
        }
    }
}

<span style="color:Blue;">public</span> <span style="color:Blue;">enum</span> ParserError
{
    MissingModifier,
    MissingBracket,
    <span style="color:Green;">// ...</span>
}

<span style="color:Green;">// Usage</span>
<span style="color:Blue;">throw</span> <span style="color:Blue;">new</span> ParserException(ParserError.MissingModifier);
</pre></div>
</div>
<p>Коду, который защищает вызов парсера почти всегда безразлично, по какой причине был завален парсинг: ему важен сам факт ошибки. Однако, если это всё-таки станет важно, пользователь всегда сможет вычленить код ошибки из свойства <code>ErrorCode</code>. Для этого вовсе не обязательно искать нужные слова по подстроке в <code>Message</code>.</p>
<p>Если отталкиваться от игнорирования вопросов переиспользования, то можно создать по типу исключения под каждую ситуацию. С одной стороны это выглядит логично: один тип ошибки - один тип исключения. Однако, тут, как и во всем, главное не переусердствовать: имея по типу исключительных операций на каждую точку выброса, вы порождаете тем самым проблемы для перехвата: код вызывающего метода будет перегружен блоками <code>catch</code>. Ведь ему надо обработать все типы исключений, которые вы хотите ему отдать. Другой минус - чисто архитектурный. Если вы не используете наследования, то тем самым дезориентируете пользователя этих исключений: между ними может быть много общего, а перехватывать их приходится по отдельности.</p>
<p>Тем не менее, существуют хорошие сценарии для введения отдельных типов для конкретных ситуаций. Например, когда поломка происходит не для всей сущности в целом, а для конкретного метода. Тогда этот тип должен быть в иерархии наследования находиться в таком месте чтобы не возникало мысли его перехватить заодно с чем-то ещё: например, выделив его через отдельную                ветвь наследования.</p>
<p>Дополнительно, если объединить эти два подхода, можно получить очень мощный инструментарий по работе с группой ошибок: можно ввести обобщающий абстрактный тип, от которого унаследовать конкретные частные ситуации. Базовый класс (наш обобщающий тип) необходимо снабдить абстрактным свойством, хранящем код ошибки, а наследники, переопределяя это свойство будут этот код ошибки уточнять:</p>
<div class="lang-csharp editor-colors"><div style="color:Black;background-color:White;"><pre>
<span style="color:Blue;">public</span> <span style="color:Blue;">abstract</span> <span style="color:Blue;">class</span> ParserException
{
    <span style="color:Blue;">public</span> <span style="color:Blue;">abstract</span> ParserError ErrorCode { <span style="color:Blue;">get</span>; }

    <span style="color:Blue;">public</span> <span style="color:Blue;">override</span> <span style="color:Blue;">string</span> Message
    {
        <span style="color:Blue;">get</span> {
            <span style="color:Blue;">return</span> Resources.GetResource($<span style="color:#A31515;">&quot;{nameof(ParserException)}{Enum.GetName(typeof(ParserError), ErrorCode)}&quot;</span>);
        }
    }
}

<span style="color:Blue;">public</span> <span style="color:Blue;">enum</span> ParserError
{
    MissingModifier,
    MissingBracket
}

<span style="color:Blue;">public</span> <span style="color:Blue;">class</span> MissingModifierParserException : ParserException
{
    <span style="color:Blue;">public</span> <span style="color:Blue;">override</span> ParserError ErrorCode { <span style="color:Blue;">get</span>; } =&gt; ParserError.MissingModifier;
}

<span style="color:Blue;">public</span> <span style="color:Blue;">class</span> MissingBracketParserException : ParserException
{
    <span style="color:Blue;">public</span> <span style="color:Blue;">override</span> ParserError ErrorCode { <span style="color:Blue;">get</span>; } =&gt; ParserError.MissingBracket;
}

<span style="color:Green;">// Usage</span>
<span style="color:Blue;">throw</span> <span style="color:Blue;">new</span> MissingModifierParserException(ParserError.MissingModifier);
</pre></div>
</div>
<p>Какие замечательные свойства мы получим при таком подходе?</p>
<ul>
<li>с одной стороны мы сохранили перехват исключения по базовому типу;</li>
<li>с другой стороны, перехватив исключение по базовому типу сохранилась возможность узнать конкретную ситуацию;</li>
<li>и плюс ко всему можно перехватить по конкретному типу, а не по базовому, не пользуясь плоской структурой классов.</li>
</ul>
<p>Как по мне так очень удобный вариант.</p>
<h3 id="section-4">По отношению к единой группе поведенческих ситуаций</h3>
<p>Какие же выводы можно сделать, основываясь на ранее описанных рассуждениях? Давайте попробуем их сформулировать:</p>
<p>Для начала давайте определимся, что имеется в виду под ситуациями. Когда мы говорим про классы и объекты, то мы привыкли в первую очередь оперировать сущностями с некоторым внутренним состоянием, над которыми можно осуществлять действия. Получается, что тем самым мы нашли первый тип поведенческой ситуации: действия над некоторой сущностью. Далее, если посмотреть на граф объектов как бы со стороны, можно заметить что он логически объединён в функциональные группы: первая занимается кэшированием, вторая - работа с базами данных, третья осуществляет математические расчёты. Через все эти функциональные группы могут идти слои: слой логирования различных внутренних состояний, журналирование процессов, трассировка вызовов методов. Слои могут быть более охватывающие: объединяющие в себе несколько функциональных групп. Например, слой модели, слой контроллеров, слой представления. Эти группы могут находиться как в одной сборке, так и в совершенно разных, но каждая из них может создавать свои исключительные ситуации.</p>
<p>Получается, что если рассуждать таким образом, то можно построить некоторую иерархию типов исключительных ситуаций, основываясь на принадлежности типа той или иной группе или слою создавая тем самым возможность перехватывающему исключения коду лёгкую смысловую навигацию в этой иерархии типов.</p>
<p>Давайте рассмотрим код:</p>
<div class="lang-csharp editor-colors"><div style="color:Black;background-color:White;"><pre>
<span style="color:Blue;">namespace</span> JetFinance
{
    <span style="color:Blue;">namespace</span> FinancialPipe
    {
        <span style="color:Blue;">namespace</span> Services
        {
            <span style="color:Blue;">namespace</span> XmlParserService
            {
            }

            <span style="color:Blue;">namespace</span> JsonCompilerService
            {
            }

            <span style="color:Blue;">namespace</span> TransactionalPostman
            {
            }
        }
    }

    <span style="color:Blue;">namespace</span> Accounting
    {
        <span style="color:Green;">/* ... */</span>
    }
}
</pre></div>
</div>
<p>На что это похоже? Как по мне, пространства имён - прекрасная возможность естественной группировки типов исключений по их поведенческим ситуациям: все, что принадлежит определённым группам там и должно находиться, включая исключения. Мало того, когда вы получите определённое исключение, то помимо названия его типа вы увидите и его пространство имён, что чётко определит его принадлежность. Помните пример плохого переиспользования типа <code>InvalidDataException</code>, который на самом деле определён в пространстве имён <code>System.IO</code>? Его принадлежность данному пространству имён означает что по сути исключение этого типа может быть выброшено из классов, находящихся в пространстве имён <code>System.IO</code> либо в более вложенном. Но само исключение при этом было выброшено совершенно из другого места, запутывая исследователя возникшей проблемы. Сосредотачивая типы исключений по тем же пространствам имён, что и типы, эти исключения выбрасывающие, вы с одной стороны сохраняете архитектуру типов консистентной, а с другой - облегчаете понимание причин произошедшего конечным разработчиком.</p>
<p>Каков второй путь группировки на уровне кода? Наследование:</p>
<div class="lang-csharp editor-colors"><div style="color:Black;background-color:White;"><pre>
<span style="color:Blue;">public</span> <span style="color:Blue;">abstract</span> <span style="color:Blue;">class</span> LoggerExceptionBase : Exception
{
    <span style="color:Blue;">protected</span> LoggerException(..);
}

<span style="color:Blue;">public</span> <span style="color:Blue;">class</span> IOLoggerException : LoggerExceptionBase
{
    <span style="color:Blue;">internal</span> IOLoggerException(..);
}

<span style="color:Blue;">public</span> <span style="color:Blue;">class</span> ConfigLoggerException : LoggerExceptionBase
{
    <span style="color:Blue;">internal</span> ConfigLoggerException(..);
}

</pre></div>
</div>
<p>Причём, если в случае с обычными сущностями приложения наследование означает наследование поведения и данных, объединяя типы по принадлежности к <em>единой группе сущностей</em>, то в случае исключений наследование означает принадлежность к <em>единой группе ситуаций</em>, поскольку суть исключения - не сущность, а проблематика.</p>
<p>Объединяя оба метода группировки, можно сделать некоторые выводы:</p>
<ul>
<li>внутри сборки (<code>Assembly</code>) должен присутствовать базовый тип исключений, которые данная сборка выбрасывает. Этот тип исключений должен находиться в корневом для сборки пространстве имён. Это будет первый слой группировки;</li>
<li>далее внутри самой сборки может быть одно или несколько различных пространств имён. Каждое из них делит сборку на некоторые функциональные зоны, тем самым определяя группы ситуаций, которые в данной сборке возникают. Это могут быть зоны контроллеров, сущностей баз данных, алгоритмов обработки данных и прочих. Для нас эти пространства имён - группировка типов по функциональной принадлежности, а с точки зрения исключений - группировка по проблемным зонам этой же сборки;</li>
<li>наследование исключений может идти только от типов в этом же пространстве имён либо в более корневом. Это гарантирует однозначное понимание ситуации конечным пользователем и отсутствие перехвата <em>левых</em> исключений при перехвате по базовому типу. Согласитесь: было бы странно получить <code>global::Finiki.Logistics.OhMyException</code>, имея <code>catch(global::Legacy.LoggerExeption exception)</code>, зато абсолютно гармонично выглядит следующий код:</li>
</ul>
<div class="lang-csharp editor-colors"><div style="color:Black;background-color:White;"><pre>
<span style="color:Blue;">namespace</span> JetFinance.FinancialPipe
{
    <span style="color:Blue;">namespace</span> Services.XmlParserService
    {
        <span style="color:Blue;">public</span> <span style="color:Blue;">class</span> XmlParserServiceException : FinancialPipeExceptionBase
        {
            <span style="color:Green;">// ..</span>
        }

        <span style="color:Blue;">public</span> <span style="color:Blue;">class</span> Parser
        {
            <span style="color:Blue;">public</span> <span style="color:Blue;">void</span> Parse(<span style="color:Blue;">string</span> input)
            {
                <span style="color:Green;">// ..</span>
            }
        }
    }

    <span style="color:Blue;">public</span> <span style="color:Blue;">abstract</span> <span style="color:Blue;">class</span> FinancialPipeExceptionBase : Exception
    {

    }
}

<span style="color:Blue;">using</span> JetFinance.FinancialPipe;
<span style="color:Blue;">using</span> JetFinance.FinancialPipe.Services.XmlParserService;

<span style="color:Blue;">var</span> parser = <span style="color:Blue;">new</span> Parser();

<span style="color:Blue;">try</span> {
    parser.Parse();
}
<span style="color:Blue;">catch</span> (XmlParserServiceException exception)
{
    <span style="color:Green;">// Something wrong in parser</span>
}
<span style="color:Blue;">catch</span> (FinancialPipeExceptionBase exception)
{
    <span style="color:Green;">// Something else wrong. Looks critical because we don&#39;t know real reason</span>
}
</pre></div>
</div>
<p>Заметьте, что тут происходит: мы как пользовательский код вызываем некий библиотечный метод, который, насколько мы знаем, может при некоторых обстоятельствах выбросить исключение <code>XmlParserServiceException</code>. И, насколько мы знаем, это исключение находится в пространстве имён, наследуя <code>JetFinance.FinancialPipe.FinancialPipeExceptionBase</code>, что говорит о возможном упущении других исключений: это сейчас микросервис <code>XmlParserService</code> создаёт только одно исключение, но в будущем могут появиться и другие. И поскольку у нас есть конвенция в создании типов исключений, мы точно знаем от кого это новое исключение будет наследоваться и заранее ставим обобщающий <code>catch</code> не затрагивая при этом ничего лишнего: то, что не попало в зону нашей ответственности пролетит мимо.</p>
<p>Как же построить иерархию типов?</p>
<ul>
<li>Для начала необходимо сделать базовый класс для домена. Назовём его доменным базовым классом. Домен в данном случае - это обобщающее некоторое количество сборок слово, которое объединяет их по некоторому глобальному признаку: логирование, бизнес-логика, UI. Т.е. максимально крупные функциональные зоны приложения;</li>
<li>Далее необходимо ввести дополнительный базовый класс для исключений, которые перехватывать необходимо: от него будут наследоваться все исключение, которые будут перехватываться ключевым словом <code>catch</code>;</li>
<li>Все исключения, которые обозначают фатальные ошибки – наследовать напрямую от доменного базового класса. Тем самым вы отделили их от перехватываемых архитектурно;</li>
<li>Разделить домен на функциональные зоны по пространствам имён и объявить базовый тип исключений, которые будут выбрасываться из каждой функциональной зоны. Тут стоит дополнительно орудовать здравым смыслом: если приложение имеет большую вложенность пространств имён, то делать по базовому типу для каждого уровня вложенности, конечно, не стоит. Однако, если на каком-то уровне вложенности происходит ветвление: одна группа исключений ушла в одно подпространство имён, а другая - в другое, то тут, конечно, стоит ввести два базовых типа для каждой подгруппы;</li>
<li>Частные исключения наследовать от типов исключений функциональных зон</li>
<li>Если группа частных исключений может быть объединена, объединить их ещё одним базовым типом: так вы упрощаете их перехват;</li>
<li>Если предполагается, что группа будет чаще перехватываться по своему базовому классу, ввести Mixed Mode c ErrorCode.</li>
</ul>
<h3 id="section-5">По источнику ошибки</h3>
<p>Ещё одним поводом для объединения исключений в некоторую группу может выступать источник ошибки. Например, если вы разрабатываете библиотеку классов, то группами источников могут стать:</p>
<ul>
<li>Вызов unsafe кода, который отработал с ошибкой. Данную ситуацию следует обработать следующим образом: обернуть исключение либо код ошибки в собственный тип исключения, а полученные данные об ошибке (например, оригинальный код ошибки) сохранить в публичном свойстве исключения;</li>
<li>Вызов кода из внешних зависимостей, который вызвал исключения, которые наша библиотека перехватить не может, т.к. они не входят в её зону ответственности. Сюда могут входить исключения из методов тех сущностей, которые были приняты в качестве параметров текущего метода или же конструктора того класса, метод которого вызвал внешнюю зависимость. Как пример, метод нашего класса вызвал метод другого класса, экземпляр которого был получен через параметры метода. Если исключение говорит о том что источником проблемы были мы сами - генерируем наше собственное исключение с сохранением оригинального - в <code>InnerExcepton</code>. Если же мы понимаем что проблема именно в работе внешней зависимости - пропускаем исключение насквозь как принадлежащее к группе внешних не подконтрольных зависимостей;</li>
<li>Наш собственный код, который был случайным образом введён в не консистентное состояние. Хорошим примером может стать парсинг текста. Внешних зависимостей нет, ухода в <code>unsafe</code> нет, а ошибка парсинга есть.</li>
</ul>

</div>
</div>
</body>
</html>